File: dia-data-import/src/main/java/ie/ncirl/diaproject/dataimport/measurement/ping/PingMeasurement.java
package ie.ncirl.diaproject.dataimport.measurement.ping;

import ie.ncirl.diaproject.dataimport.measurement.Measurement;

public class PingMeasurement extends Measurement {

    public Task task;
    public Parameters parameters;
    public String success;
    public String timestamp;
    public DeviceProperties device_properties;
    public Values values;
    public String type;
    public String id;

    @Override
    public String toHdr(String sep) {
        StringBuffer sb = new StringBuffer();
        separate(sb, (new Task()).toHdr(sep), sep);
        separate(sb, (new Parameters()).toHdr(sep), sep);
        separate(sb, "m_success", sep);
        separate(sb, "m_timestamp", sep);
        separate(sb, (new DeviceProperties()).toHdr(sep), sep);
        separate(sb, (new Values()).toHdr(sep), sep);
        separate(sb, "m_type", sep);
        separate(sb, "m_id", NO_SEP);
        return(sb.toString());
    }

    @Override
    public String toCsv(String quote, String sep) throws NullPointerException {
        StringBuffer sb = new StringBuffer();

        separate(sb, task != null
                ? task.toCsv(quote, sep)
                : (new Task()).toNullCsv(sep), sep);

        separate(sb, parameters != null
                ? parameters.toCsv(quote, sep)
                : (new Parameters()).toNullCsv(sep), sep);

        quoteAndSeparate(sb, success, quote, sep);
        quoteAndSeparate(sb, timestamp, quote, sep);

        separate(sb, device_properties != null
                ? device_properties.toCsv(quote, sep)
                : (new DeviceProperties()).toNullCsv(sep), sep);

        separate(sb, values != null
                ? values.toCsv(quote, sep)
                : (new Values()).toNullCsv(sep), sep);

        quoteAndSeparate(sb, type, quote, sep);
        quoteAndSeparate(sb, id, quote, NO_SEP);
        return(sb.toString());
    }

    @Override
    public String toNullCsv(String sep) {
        StringBuffer sb = new StringBuffer()
                .append((new Task()).toNullCsv(sep)).append(sep) // task
                .append((new TaskParameters()).toNullCsv(sep)).append(sep) // parameters
                .append(sep) // success
                .append(sep) // timestamp
                .append((new DeviceProperties()).toNullCsv(sep)).append(sep) // device_properties
                .append(sep) // values
                .append(sep) // type
                .append(NO_SEP); // id
        return(sb.toString());
    }
}

File: dia-spark-app/src/main/scala/ie/ncirl/diaproject/dataprocess/measurement/ping/PingMeasurement.scala
package ie.ncirl.diaproject.dataprocess.measurement.ping

import ie.ncirl.diaproject.dataprocess.measurement.DeviceProperties

case class PingMeasurement
(
  task: Task,
  parameters: Parameters,
  success: String,
  timestamp: String,
  device_properties: DeviceProperties,
  values: Values,
  `type`: String,
  id: String
)

case class Task
(
  count: String,
  filter: String,
  parameters: TaskParameters,
  created: String,
  start_time: String,
  interval_sec: String,
  priority: String,
  tag: String,
  end_time: String,
  `type`: String,
  id: String
)

case class TaskParameters
(
  location_update_distance: String,
  target: String,
  trigger_location_update: String,
  packet_size_byte: String,
  ping_timeout_sec: String,
  profile_1_freq: String,
  profile_2_freq: String,
  profile_3_freq: String,
  profile_4_freq: String,
  profile_unlimited: String
)

case class Parameters
(
  count: String,
  target: String,
  parameters: String,
  start_time: String,
  interval_sec: String,
  ping_timeout_sec: String,
  priority: String,
  packet_size_byte: String,
  ping_exe: String,
  key: String,
  `type`: String,
  end_time: String,
  myspeedtest: String,
  ping: String,
  options: String,
  context_interval_sec: String,
  ping_time_to_live: String,
  ping_icmp_interval_sec: String,
  sensitive: String,
  ping_method: String,
  use_https: String,
  port: String
)

case class Values
(
  packets_sent: Integer,
  packet_loss: Integer,
  mean_rtt_ms: Integer,
  CalledByLocation: String,
  MeasurementLatitude: String,
  max_rtt_ms: Integer,
  target_ip: String,
  stddev_rtt_ms: Integer,
  min_rtt_ms: Integer,
  MeasurementLongitude: String,
  filtered_mean_rtt_ms: Integer,
  error: String,
  ping_method: String,
  battery: String,
  src_ip: String,
  dst_ip: String,
  measure: String,
  time: String,
  context_results: String,
  mptcp_config: String
)
